### ISSUE TITLE
Dependency pinning in 0.9.24 does not seem to be working

### ISSUE THREAD

**Issue by howbazaar** [cloud-custodian/cloud-custodian#8429]
### Describe the bug

Installing via poetry does not use pinned versions, whereas installing via pip does.

In a fresh poetry environment, I did the following:

```shell
poetry add c7n=^0.9.24
```

This installed the following dependencies:

```
  • Installing six (1.16.0)
  • Installing jmespath (1.0.1)
  • Installing python-dateutil (2.8.2)
  • Installing urllib3 (1.26.15)
  • Installing botocore (1.29.98)
  • Installing attrs (22.2.0)
  • Installing pyrsistent (0.19.3)
  • Installing s3transfer (0.6.0)
  • Installing zipp (3.15.0)
  • Installing argcomplete (3.0.4)
  • Installing boto3 (1.26.98)
  • Installing docutils (0.18.1)
  • Installing importlib-metadata (5.2.0)
  • Installing jsonschema (4.17.3)
  • Installing pyyaml (6.0)
  • Installing tabulate (0.9.0)
  • Installing c7n (0.9.24)
```

I expect to be able to export from poetry to a requirements file and use that in a fresh virtual environment.

```export
poetry export --without-hashes -o requirements.txt
poetry run python -m venv test-env 
. test-env/bin/activate
pip install -r requirements.txt 
```

However this currently errors out with conflicting dependencies.

### What did you expect to happen?

I expected the virtual environment to be able to install the versions specified in the `requirements.txt` file.

### Cloud Provider

Amazon Web Services (AWS)

**Comment by kapilt** [comment]
its odd, I was able to verify the wheel itself is frozen in terms of downloading the content from pypi and examining its metadata record

```
<ZipInfo filename='c7n-0.9.24.dist-info/METADATA' compress_type=deflate filemode='?rw-r--r--' file_size=13518 compress_size=5000>                                                                                                                  
>>> f.open(name=f.filelist[-2])                                                                                                                                                                                                                    
<zipfile.ZipExtFile name='c7n-0.9.24.dist-info/METADATA' mode='r' compress_type=deflate>                                                                                                                                                           
>>> y = f.open(name=f.filelist[-2]) 
>>> from email import parser 
>>> m = parser.Parser().parsestr(contents.decode('utf8')) 
>>> m.get_all('Requires-Dist')
['argcomplete (==2.1.1)', 'boto3 (==1.26.86)', 'docutils (==0.18.1)', 'importlib-metadata (==5.2.0)', 'jsonschema (==4.17.3)', 'python-dateutil (==2.8.2)', 'pyyaml (==6.0)', 'tabulate (==0.9.0)']

**Comment by kapilt** [comment]
its odd, I was able to verify the wheel itself is frozen in terms of downloading the content from pypi and examining its metadata record

```
<ZipInfo filename='c7n-0.9.24.dist-info/METADATA' compress_type=deflate filemode='?rw-r--r--' file_size=13518 compress_size=5000>                                                                                                                  
>>> f.open(name=f.filelist[-2])                                                                                                                                                                                                                    
<zipfile.ZipExtFile name='c7n-0.9.24.dist-info/METADATA' mode='r' compress_type=deflate>                                                                                                                                                           
>>> y = f.open(name=f.filelist[-2]) 
>>> from email import parser 
>>> m = parser.Parser().parsestr(contents.decode('utf8')) 
>>> m.get_all('Requires-Dist')
['argcomplete (==2.1.1)', 'boto3 (==1.26.86)', 'docutils (==0.18.1)', 'importlib-metadata (==5.2.0)', 'jsonschema (==4.17.3)', 'python-dateutil (==2.8.2)', 'pyyaml (==6.0)', 'tabulate (==0.9.0)']
```

**Comment by kapilt** [comment]
its odd, I was able to verify the wheel itself is frozen in terms of downloading the content from pypi and examining its metadata record

```
<ZipInfo filename='c7n-0.9.24.dist-info/METADATA' compress_type=deflate filemode='?rw-r--r--' file_size=13518 compress_size=5000>                                                                                                                  
>>> f.open(name=f.filelist[-2])                                                                                                                                                                                                                    
<zipfile.ZipExtFile name='c7n-0.9.24.dist-info/METADATA' mode='r' compress_type=deflate>                                                                                                                                                           
>>> y = f.open(name=f.filelist[-2]) 
>>> from email import parser 
>>> m = parser.Parser().parsestr(contents.decode('utf8')) 
>>> m.get_all('Requires-Dist')
['argcomplete (==2.1.1)', 'boto3 (==1.26.86)', 'docutils (==0.18.1)', 'importlib-metadata (==5.2.0)', 'jsonschema (==4.17.3)', 'python-dateutil (==2.8.2)', 'pyyaml (==6.0)', 'tabulate (==0.9.0)']
```

**Comment by kapilt** [comment]
its odd, I was able to verify the wheel itself is frozen in terms of downloading the content from pypi and examining its metadata record

```
<ZipInfo filename='c7n-0.9.24.dist-info/METADATA' compress_type=deflate filemode='?rw-r--r--' file_size=13518 compress_size=5000>                                                                                                                  
>>> f.open(name=f.filelist[-2])                                                                                                                                                                                                                    
<zipfile.ZipExtFile name='c7n-0.9.24.dist-info/METADATA' mode='r' compress_type=deflate>                                                                                                                                                           
>>> y = f.open(name=f.filelist[-2]) 
>>> from email import parser 
>>> m = parser.Parser().parsestr(contents.decode('utf8')) 
>>> m.get_all('Requires-Dist')
['argcomplete (==2.1.1)', 'boto3 (==1.26.86)', 'docutils (==0.18.1)', 'importlib-metadata (==5.2.0)', 'jsonschema (==4.17.3)', 'python-dateutil (==2.8.2)', 'pyyaml (==6.0)', 'tabulate (==0.9.0)']
```

**Comment by wallabyies** [comment]
Somewhat related, if installing a pinned version of Custodian via pip, how are the requirements done without the requirements file now?

**Comment by kapilt** [comment]
sorry for the extra copies of comments, seems an artifact of GitHub going awol earlier.
https://www.githubstatus.com/incidents/52z0j6phhnjs


@wallabyies the wheel itself has pinned metadata, that's been the same between releases, we now use https://github.com/cloud-custodian/poetry-plugin-freeze to generate those frozen wheels instead of generating setup.py and requirements.txt artifacts

@howbazaar I was able to track this down to a delta in how poetry resolves versus pip and also how poetry publishes to pypi. basically for resolution on dependencies poetry is using the simple json api on pypi, instead of wheel metadata (which is understandable, its faster vs downloading and parsing the wheel metadata). the issue is that on publishing its using the pyproject.toml contents instead of the wheel metadata to register metadata about the package version, so its ignoring the frozen dependency specification in the wheel. I think we'll need to switch back to twine for publishing to resolve that. 
filed upstream in poetry as https://github.com/python-poetry/poetry/issues/7723

for now we'll need a new release to resolve.

**Comment by kapilt** [comment]
@wallabyies re requirements pinning you can just pin custodian and provider plugins, so c7n==0.9.23 suffices to get a frozen dep set, alternatively installing and pip freeze also works (and is the common usage)

@howbazaar fwiw, some additional comparison of freeze plugin versus previous frozen releases revealed some another minor anomalies (only direct deps were frozen instead of full graph), that's resolved in 1.0.3 of the plugin per https://github.com/cloud-custodian/poetry-plugin-freeze/pull/10
